src/main/io/gps.c

```diff
@@ -573,110 +759,388 @@ static void ubloxSetMessageRate(uint8_t messageClass, uint8_t messageID, uint8_t
+    } else if (gpsData.ubloxM8orAbove) {
+        ubloxSetMessageRate(CLASS_NAV, MSG_NAV_SAT, divisor);
+    } else {
+        ubloxSetMessageRate(CLASS_NAV, MSG_NAV_SVINFO, divisor);
     }
+}
 
-    tx_buffer.payload.cfg_sbas.maxSBAS = 3;
-    tx_buffer.payload.cfg_sbas.scanmode2 = 0;
-    switch (gpsConfig()->sbasMode) {
-        case SBAS_AUTO:
-            tx_buffer.payload.cfg_sbas.scanmode1 = 0;
-            break;
-        case SBAS_EGNOS:
-            tx_buffer.payload.cfg_sbas.scanmode1 = 0x00010048; //PRN123, PRN126, PRN136
-            break;
-        case SBAS_WAAS:
-            tx_buffer.payload.cfg_sbas.scanmode1 = 0x0004A800; //PRN131, PRN133, PRN135, PRN138
-            break;
-        case SBAS_MSAS:
-            tx_buffer.payload.cfg_sbas.scanmode1 = 0x00020200; //PRN129, PRN137
-            break;
-        case SBAS_GAGAN:
-            tx_buffer.payload.cfg_sbas.scanmode1 = 0x00001180; //PRN127, PRN128, PRN132
-            break;
-        default:
-            tx_buffer.payload.cfg_sbas.scanmode1 = 0;
+#endif // USE_GPS_UBLOX
+
+#ifdef USE_GPS_NMEA
+void gpsInitNmea(void)
+{
+    static bool atgmRestartDone = false;
+#if !defined(GPS_NMEA_TX_ONLY)
+    uint32_t now;
+#endif
+    DEBUG_SET(DEBUG_GPS_CONNECTION, 4, gpsData.state);
+    DEBUG_SET(DEBUG_GPS_CONNECTION, 5, gpsData.state_position);
+    switch (gpsData.state) {
+        case GPS_STATE_INITIALIZING:
+#if !defined(GPS_NMEA_TX_ONLY)
+           now = millis();
+           if (cmp32(now, gpsData.state_ts) < 1000) {
+               return;
+           }
+           gpsData.state_ts = now;
+            switch (gpsData.state_position) {
+                case 0: // first run after bootup
+                    serialSetBaudRate(gpsPort, 4800);
+                    gpsData.state_position++;
+                    break;
+                case 1: // second run
+                    // print the init string for the baudrate we want to be at
+                    serialPrint(gpsPort, "$PSRF100,1,115200,8,1,0*05\r\n");
+                    gpsData.state_position++;
+                    break;
+                default: 
+                    // we're now (hopefully) at the correct rate, next state should switch to it
+                    // TO DO: Check that this actually works, it seems broken at present
+                    gpsData.state_position = 0;
+                    gpsSetState(GPS_STATE_CHANGE_BAUD);
+                    break;
+            }
             break;
+#endif
+        case GPS_STATE_CHANGE_BAUD:
+#if !defined(GPS_NMEA_TX_ONLY)
+            // wait a short time between sending commands
+            // note that no commands are sent to request the packets we need
+            now = millis();
+            if (cmp32(now, gpsData.state_ts) < 500) {
+                return;
+            }
+            gpsData.state_ts = now;
+
+            if (gpsData.state_position < 1) { // first run after boot up 
```

Comment:
(switch(gpsData.state_position))

Reply:
it would be nice if this was refactored to switch statements, but could I please leave that for another PR?